### ISSUE TITLE
Options to Select the "Best" Hyperparameter

### ISSUE THREAD

**Issue by azev77** [alan-turing-institute/MLJ.jl#487]
Suppose I split a data set into: XT/YT (training) & XH/YH (holdout) 
I train a Lasso model w/ one HP: lambda in HP_Grid. 
1. For each lambda, I train the model within subsets of XT/YT, using my favorite **resampling procedure**  (K fold CV etc). 
2. For each lambda, I compute my favorite **resampling score** (RMSE etc) within folds of XT/YT not used for training. (This is an out-of sample score w/i the training data.)
Now, for each lambda, I have a score. 
Q: how do I select the "optimal hyper-parameter" 
3. A: the "usual approach" is to select the lambda with the **minimum score** (minimum error/maximum R2...)
B: Tibshirani & co warn this may not be prudent!
Instead they recommend: select lambda 1 standard error (se) above the minimum CV error.
Often, there may be multiple such lambda, they recommend the **most parsimonious** such model. In this context most parsimonious = most regularized = biggest lambda.   
In all-subset regression parsimonious = model w/ fewest non-zero coef. 

Bottom line: a flexible ML interface gives the user the ability to choose her favorite model selection procedure, given her objectives

@ryantibs [illustrates this nicely](https://www.stat.cmu.edu/~ryantibs/datamining/lectures/19-val2.pdf) in his slides:
![image](https://user-images.githubusercontent.com/7883904/78862234-de85f400-7a04-11ea-8e5d-a3b64605d53b.png)
Here is how I do this in R:
```R
  #Train Lasso
  m= cv.glmnet(XT, YT,alpha=1, nfolds=nf, intercept=T) 
  # select HP w/ minimum CV error
  pr= predict(m, newx=XH, s="lambda.min")
  os= sc(YH, pr)   #RMSE 
  # select HP w/ one se rule
  pr= predict(m, newx=XH, s="lambda.1se")
  os= sc(YH, pr)   #RMSE 
```
Next train the model on the FULL XT/YT using preferred lambda.
Compute the score on XH/YH. 
Often "lambda.1se" outperforms "lambda.min" out-of-sample in the Holdout data.

Parsimony a bit subjective, because there are different notions of parsimony.
There can be a standard default option for each model & enough flexibility for the user to define her own procedure.

**Comment by ablaom** [comment]
@azev77 Thanks for this. This sounds like a great idea. 

It has been suggested to me privately that something like the above be implemented by adding a `model_selection_strategy' field to TunedModel, which would allow user to choose, say, the "one-standard error" model or the "optimal" model. 

This makes sense but requires that we know how to extract a score from each `result` object. Currently, the history is a collection of  `(model, result)` pairs (where `model` = hyperparameter set) and  the format of `result` is left up to the tuning-strategy. The `result` will generally contain *multiple* scores because: (i) the user may be interested in metrics beyond the one defining the objective, and (ii) we want to include multi-objective (Pareto) optimisation strategies. Also, the result may contain strategy-specific data. For example the Tree Parzen optimisation strategy coming soon needs to record a custom representation of the model there. 

Currently `result` is required to be a named tuple, and in the case of the 3 built-in tuning strategies **there is always a key called `measurement` which is a vector of one or more scores, the first corresponding to the metric defining the objective function**. So we could simply make that a firm requirement, yes?

@yalwan-iqvia Can you confirm the above statement holds for Tree Parzen?

@azev77 

> Parsimony a bit subjective, because there are different notions of parsimony.
> There can be a standard default option for each model

Yes, in principle. In the meantime, there needs to be a fallback option, independent of model. Can you think of something better than random selection?

> enough flexibility for the user to define her own procedure.

Maybe `model_selection_stragegy` could also be a callable object with signature

```julia
model_selection_strategy(history)
```
and we document how to extract what is needed from the the history? (basically what I did above). Any other suggestion?

**Comment by azev77** [comment]
Minor point: `model_selection_strategy` can mean different things. 
1. Select between multiple different models {Flux, Lasso, Ridge}
2. Select the optimal HP vector for a given model {optimal lambda for Lasso}

Maybe `HP_selection_strategy` but that sounds weird. 

In general, `HP_selection_strategy` **is a function of (at least) two variables**: `HP_Grid` & `Score`. 
Eg: for Lasso, we have a grid of lambdas & a vector of "scores" for each lambda

The default `HP_selection_strategy`: returns the HP vector that minimizes the score
```Julia
f(HP_Grid, Score)= HP_Grid[ argmin[Score] ]
```
Note: this is sometimes min (RMSE), sometimes max (Accuracy) 
Not sure I like the phrase `best` or `optimal` bc it is only best/optimal in some context...

The second most common `HP_selection_strategy`: returns the HP vector 1 se above the one that minimizes the score. 
(1 se below the one that maximizes the score)
Note: below I will make the number of se a keyword arg w/ default =1
```Julia
function f(HP_Grid, Score; n_se=1)
        se= StandardError(Score)
        m= min[Score] + n_se * se
        ix= #Index of Score closest to m
        hp= HP_Grid[ ix ]
        return hp
end
```
Tibshirani & co view regularized as a synonym for "parsimonious" & select the biggest lambda w/ score closest to m. 
{That probably makes sense.} 

Finally, a user can define her own favorite `HP_selection_strategy`:
```Julia
function f(HP_Grid, Score; \theta =1)
        ...
        return hp
end
```

**Comment by ablaom** [comment]
> In general, HP_selection_strategy is a function of (at least) two variables: HP_Grid & Score

The history contains both the models (hyperparameter values, or "grid" in your sense, I guess) and the corresponding scores. See the readme of this repo for a detailed description of the existing API, which furthermore explains all terminology I am using in the post above. 

> Note: this is sometimes min (RMSE), sometimes max (Accuracy)

At present *any* MLJ supported measure or measures can be specified when constructing a `TunedModel` instance. See the readme.

> Finally, a user can define her own favorite HP_selection_strategy

Yes, this is more-or-less what I am suggesting, no?

**Comment by azev77** [comment]
Yes it is what your suggesting. I tend to be pedantic. 
I just looked @: https://github.com/alan-turing-institute/MLJTuning.jl
I see history returns tuple `(m , r)` since `m` contains the grid & `r` contains scores we're in business...

**Comment by yalwan-iqvia** [comment]
> @yalwan-iqvia Can you confirm the above statement holds for Tree Parzen?

Actually, yes, I worked (quite) hard to design around the existing design -- one of this was about being able to not have to implement `best` ourselves, and so we ensured that the `result` function returned as similar as possible to fallback version, which, yes, includes the `measurement` key.

**Comment by yalwan-iqvia** [comment]
Also additional remark, I **think** we would be happy to accommodate breaking changes.

**Comment by ablaom** [comment]
@yalwan-iqvia Thanks for the clarification. Unfortunately, I think you can expect breaking changes in most 0.x open-source software projects 😄, and we're only at 0.2.x.

Let's sit on this a wee bit before committing to anything. 

@azev77 Although one can distinguish between models that come from the same algorithm and models the differ only in the values of hyperparameters, some model selection criterion might apply in the greater generality and be useful in the specific case here, no? If so, there could be overlap with what we do here and the benchmark design. What do you think?

**Comment by ablaom** [comment]
> not all resampling strategies have these "folds".

Sounds interesting. In MLJ, the resampling API only allows implementation of resampling strategies that generate a vector of `(train, test)` pairs, where `train` and `test` are subsets of a set of provided indices, `rows` for the data. There's no restriction on how these interact (overlap, and so forth). If you have something that does not fit in here, please open an issue at MLJBase.jl with a use- case. (You mentioned "Monte Carlo resampling" somewhere, but given the breadth of application of these terms, I probably need a bit of detail or a reference, thanks!).

**Comment by yalwan-iqvia** [comment]
This stack exchange post does a lot of the hard work for me here:
https://stats.stackexchange.com/questions/51416/k-fold-vs-monte-carlo-cross-validation

An answerer points out that sklearn seems to have this, in a multitude of flavours. no less, as well.

**Comment by ablaom** [comment]
Looks to me like this fits into the existing framework https://alan-turing-institute.github.io/MLJ.jl/dev/evaluating_model_performance/#Custom-resampling-strategies-1 just fine. You're still generating a bunch of (train, test) sets of indices. It just so happens that there is potential overlap between individual trains and individual tests.

Am I missing something here?

**Comment by ablaom** [comment]
If you wanted the folds lazily generated, that might require some changes to the API, but I'd say that's probably an unnecessary optimisation.

**Comment by ablaom** [comment]
Okay, I think Monte-Carlo is already possible; see https://github.com/alan-turing-institute/MLJ.jl/issues/564#issuecomment-642932853

**Comment by yalwan-iqvia** [comment]
My point about folds there was purely the naming as opposed to something else, i.e. I think `per_fold` should be called `per_resample`

**Comment by ablaom** [comment]
> My point about folds there was purely the naming as opposed to something else, i.e. I think `per_fold` should be called `per_resample`

Yes, I agree that `per_resample` is better than `per_fold`. However, given the name change would be breaking, I'm disinclined to change it.
